### PR TITLE
ARROW-5115: [JS] Add Vector Builders and high-level stream primitives

### LINKED ISSUE #4452: Arrow TS Excessively Deep Type Instantiations
Using the following versions of arrow and typescript:

- `@apache-arrow/ts`: 0.13.0
- `typescript`: 3.5.1

I get a slew of errors like the following on compilation:

```
ERROR in [at-loader] ../../node_modules/@apache-arrow/ts/vector/base.ts:69:16 
    TS2589: Type instantiation is excessively deep and possibly infinite.
```

I am new to Arrow and relatively new to Typescript, so please let me know any additional information that would be helpful and I am happy to provide. Thanks!

### PR DESCRIPTION
**Pull request by trxcllnt**
This PR adds Vector Builder implementations for each DataType, as well as high-level stream primitives for Iterables/AsyncIterables, node streams, and DOM streams.

edit: I've created a demo that transforms a CSV file/stream of JSON rows to an Arrow table in this repository: https://github.com/trxcllnt/csv-to-arrow-js

#### Builder API

The new `Builder` class exposes an API for sequentially appending (or setting into slots that have already been allocated) arbitrary JavaScript values that will be flushed to the same underlying Data chunk. The `Builder` class also supports specifying a list of null-value sentinels, or values that will be interpreted to indicate "null" should be written to the null bitmap instead of being written as a valid element.

Similar to the existing `Vector` API, `Builder` has a static `Builder.new()` method that will return the correct `Builder` subclass instance for the supplied DataType. Since the `Builder` constructor takes an options Object, this method also takes an Object:

```typescript
import { Builder, Utf8 } from 'apache-arrow';

const utf8Builder = Builder.new({
    type: new Utf8(),
    nullValues: [null, 'n/a']
});

utf8Builder
    .append('hello')
    .append('n/a') // will be interpreted to mean `null`
    .append('world')
    .append(null);

const utf8Vector = utf8Builder.finish().toVector();

console.log(utf8Vector.toJSON());
// > ["hello", null, "world", null]
```

The `Builder` class has two methods for flushing the pending values to their underlying ArrayBuffer representations: `flush(): Data<T>` and `toVector(): Vector<T>` (`toVector()` calls `flush()` and creates a `Vector` instance from the returned Data instance).

Calling `Builder.prototype.finish()` will finalize a `Builder` instance. After this, no more values should be written to the Builder instance. This is a no-op on for most types, except the `DictionaryBuilder`, which flushes its internal dictionary and writes the values to the `Dictionary` type's `dictionaryVector` field.

#### Iterable and stream APIs

Creating and using Builders directly is a bit cumbersome, so we provide some high-level streaming APIs for automatically creating builders, appending values, and flushing chunks of a certain size:

```typescript
Builder.throughIterable(options: IterableBuilderOptions<T, TNull>)
Builder.throughAsyncIterable(options: IterableBuilderOptions<T, TNull>)
Builder.throughDOM(options: BuilderTransformOptions<T, TNull>)
Builder.throughNode(options: BuilderDuplexOptions<T, TNull>)
```

#### Iterables and AsyncIterables
The static `throughIterable` and `throughAsyncIterable` methods take an `options` argument that indicates the Builder's type and null-value sentinels, and returns a function which accepts an Iterable or AsyncIterable, respectively, of values to transform:

```typescript
import { Chunked, Builder, Utf8 } from 'apache-arrow';
const options = { type: new Utf8(), nullValues: [null, 'n/a'] };
const buildUtf8 = Builder.throughIterable(options);
const utf8Vector = Chunked.concat(...buildUtf8(['hello', 'n/a', 'world', null]));
```

The `options` argument can also specify a `queueingStrategy` and `highWaterMark` that control the chunking semantics:
* If the `queueingStrategy` is `"count"` (or is omitted), then the returned generator function will flush the `Builder` and yield a chunk once the number of values that have been written to the Builder reaches the value supplied for `highWaterMark`, regardless of how much memory the `Builder` has allocated.
* If the `queueingStrategy` is `"bytes"`, then the returned generator function will flush the `Builder` and yield a new chunk once the Builder's `byteLength` field reaches or exceeds the value supplied for `highWaterMark`, regardless of how many elements the `Builder` contains.

#### Node and DOM Streams

In addition to the Iterable transform APIs, we can also create node and DOM transform streams with similar options:
```typescript
import { Readable } from 'stream';
import { toArray } from 'ix/asynciterable/toarray';
import { Chunked, Builder, Utf8 } from 'apache-arrow';
const options = {
  type: new Utf8(),
  nullValues: [undefined, 'n/a'],
  queueingStrategy: 'bytes',
  highWaterMark: 64, // flush each chunk once 64 bytes have been written
};

const utf8Vector = Chunked.concat(await toArray(
  Readable
    .from(['hello', 'n/a', 'world', undefined])
    .pipe(Builder.throughNode(options))
));
```

#### Miscellaneous
* Updates most dependencies, updates TypeScript to v3.5.1 (and resolves #4452)
* Updates the BigInt compatibility type to use `Object.setPrototypeOf()`, yielding a 4x speedup
* Updates Int64 and Uint64 set routines to accept native `bigint` types if available
* Adds a docstring to the `_InternalEmptyPlaceholderRecordBatch` class added in #4373

### CODE REVIEW

**Comment by wesm** [comment]
This is a pretty epic piece of work, @TheNeuralBit @domoritz I assume this is on your radar to review?

File: js/package.json
Comment (domoritz) [inline js/package.json]: At some pint, we should replace tslint with slint since the former is being deprecated.

**Review by domoritz — COMMENTED** [review]
This is way too much code for me to review right now, unfortunately.

File: js/package.json
Comment (trxcllnt) [reply js/package.json]: yes I imagine at some point we should switch over.

**Review by TheNeuralBit — COMMENTED** [review]
I'm in the same boat as Dom, not sure I can devote enough time to review this thoroughly. Maybe we should schedule some time to just walk through it together.

File: js/package.json
Comment (trxcllnt) [reply js/package.json]: Yes unfortunately :/. Each TypeDoc release is pinned to a specific version of the TypeScript compiler, and as of this commit, the latest public release on npm is on TS 3.2.4, but master is on 3.4.5. We should update this dependency when TypeDoc publishes master to npm.

File: js/test/unit/builders/builder-tests.ts
Comment (trxcllnt) [reply js/test/unit/builders/builder-tests.ts]: Ah yes it is. I had this loop in while I was debugging an edge case and forgot to remove it once it was solved.

File: js/package.json
Comment (trxcllnt) [reply js/package.json]: The changes to update to TS@3.5.1 were relatively minor changes to the type mapping interfaces, and are all in this commit: https://github.com/trxcllnt/arrow/commit/492d1f111bb78b75ca79790beb4654119af93d51#diff-3cc15e0c860bac3718152b2c143638ad

I don't disagree on preferring this to be split out and handled separately, but practically speaking I was working under a number of constraints that made it more difficult. First, TypeScript released v3.3+ with breaking changes to type resolution halfway through the process of working on these features (leading to [this issue]( https://github.com/apache/arrow/issues/4452)).

Second, I was using the new Builders in app code as they were being developed. The apps needed TS 3.5, but wouldn't compile without these fixes, so I had to update Arrow. But the Builders branch was sufficiently different and would have been difficult to track and constantly rebase with that change (and wasn't even done yet), it was easier to do it here.

File: js/src/builder/dictionary.ts
Comment (trxcllnt) [reply js/src/builder/dictionary.ts]: Ha yeah, this is an interesting bit of code. I pulled it from a collection of neat hash functions I've seen over the years, which I realize now might not be a good idea. I'd initially tried to use DJB2 which is fast and simple 32-bit hash, and as you point out, is subject to collisions. So this one, while not as fast, is a 53-bit hash and should have a lower likelihood of collisions.

Re: terminology, I've been doing a bunch of python lately, and I think I used the same names as pandas' Categorical dtype :sweat_smile:.

As to why I'm hashing at all, JS strings take up an enormous amount of memory. That memory is allocated on the heap, is subject to the v8 memory limits, and can easily grow to be the majority of VM-allocated memory in an app. But when those strings are encoded to Uint8Arrays (and the string is GC'd), the Uint8Arrays _aren't_ subject to the same v8 memory limits. So by hashing each string to a number, we reduce the likelihood of OOM'ing the JS VM accidentally.

In node I've been using the [`node-metrohash`](https://github.com/robertklep/node-metrohash/) native module for fast hashing, but the browser doesn't have any similar fast hashing APIs (there's a cryptography API, but it's promise based and cryptographically secure, so not great for hashing). I wanted a reasonable default if someone wants to dictionary-encode some data but doesn't provide their own hash function.

**Comment by trxcllnt** [comment]
@TheNeuralBit that sounds good to me, I'm happy to schedule some time to do a walk through whenever you're free.

I tried to model the Builders similar to the Vectors, both organizationally and in their public API. Most of the additions live in the [`src/builders`](https://github.com/trxcllnt/arrow/tree/js/data-builders/js/src/builder) directory, or the [`src/Builder.ts` base class](https://github.com/trxcllnt/arrow/blob/js/data-builders/js/src/builder.ts), which I tried to annotate with docstrings as I went. I also followed a similar pattern to the Arrow C++ Builders, and centralized the low-level buffer operations into a few [`BufferBuilder` classes here](https://github.com/trxcllnt/arrow/blob/js/data-builders/js/src/builder/buffer.ts).

Creating and using Builders is also very similar to Vectors: `Builder.new({ type: DataType })` returns a new Builder for the given `type`, and you can `set()` or `append()` any of the same types that are accepted or returned by the corresponding `Vector.set()` and `Vector.get()` calls, respectively. The high-level streaming APIs also accept the same options as `Builder.new()` to help make the transition from an imperative building routine to a streaming routine as easy as possible.

And lastly, I added a basic implementation for `Vector.from()` for all Vectors that use the corresponding Builder implementation to create a Vector from JS values for simple use-cases (and Table/RecordBatch that use the Struct or Map Builders under the hood). For example, now you can create a Table from JS objects rather easily:

```typescript
const { Table, Field, Map_, Int32, Utf8 } = require('apache-arrow');
const table = Table.from({
  type: new Map_([new Field('int', new Int32()), new Field('str', new Utf8())]),
  values: [
    { 'int': 0, str: 'hello' },
    { 'int': 1, str: 'world' }
  ]
});
console.table([...table].map((row) => row.toJSON());
```

**Comment by bclinkinbeard** [comment]
Just a note that my team is looking forward to this feature being available. Our backend serves binary protocol buffers with custom bit packing strategies, so we do a fair amount of dynamic Arrow table construction. I expect Builders will allow us to remove some pretty repetitive code.

**Comment by wesm** [comment]
@trxcllnt @TheNeuralBit think we can get this merged this week? Let me know if there's something I can do to help

**Comment by trxcllnt** [comment]
@wesm yes, I'd love to get this (and the delta dictionaries PR https://github.com/apache/arrow/pull/4502) merged this week for the 0.14 release. I've been using them in side projects and they're pretty stable, I think we're just blocked on a review here. @TheNeuralBit let me know if you have time to schedule a walkthrough some time this week/weekend, or @wesm I'd also be happy to do one with you.

**Comment by TheNeuralBit** [comment]
Paul and I are meeting up today to go over this. Merging this week seems reasonable.

File: js/src/builder/dictionary.ts
Comment (TheNeuralBit) [reply js/src/builder/dictionary.ts]: As discussed, can you add a commit that removes this and makes the default hash function identity?

We don't have any license for this code so we shouldn't commit it

File: js/src/builder/valid.ts
Comment (TheNeuralBit) [inline js/src/builder/valid.ts]: Could you add an example of what the generated function should look like here? I find that pretty helpful when looking through code and encountering some kind of code generation.

**Review by TheNeuralBit — CHANGES_REQUESTED** [review]
Thanks for walking me through this, and thanks for implementing it. It's a great addition and I think it will make it a lot easier for people to start using arrow JS.

Just a couple of minor comments based on our discussion.

**Review by TheNeuralBit — APPROVED** [review]

**Comment by TheNeuralBit** [comment]
Thanks! I'll merge when CI passes

**Pull request closed without merging**
